# **1. Introdução e Contexto**

### **Tech Challenge - Modelo Preditivo de Diagnóstico de Obesidade**

**Objetivo:** Desenvolver um modelo de Machine Learning capaz de prever o nível de obesidade ou sobrepeso de um indivíduo com base em suas características demográficas, físicas e hábitos comportamentais para auxiliar a tomada de decisão da equipe médica.
**Meta Técnica:** Assertividade (Acurácia) mínima acima de 75% no conjunto de teste e deploy da aplicação preditiva em produção.

Dados extraídos do prontuário/questionário hospitalar: `Obesity.csv`

#### **Informações importantes sobre a base de dados**

* **Gender:** Gênero do paciente.
* **Age:** Idade.
* **Height:** Altura em metros.
* **Weight:** Peso em quilogramas (kgs).
* **family_history:** Histórico familiar (algum membro sofreu ou sofre de excesso de peso?).
* **FAVC:** Consumo frequente de alimentos altamente calóricos.
* **FCVC:** Frequência de consumo de vegetais nas refeições.
* **NCP:** Número de refeições principais feitas diariamente.
* **CAEC:** Consumo de alimentos entre as refeições.
* **SMOKE:** Paciente possui hábito de fumar?
* **CH2O:** Quantidade de água consumida diariamente.
* **SCC:** Monitoramento diário das calorias ingeridas.
* **FAF:** Frequência de prática de atividade física.
* **TUE:** Tempo diário de uso de dispositivos tecnológicos (celular, computador, videogame, etc.).
* **CALC:** Frequência de consumo de álcool.
* **MTRANS:** Meio de transporte mais utilizado no cotidiano.
* **Obesity (Variável Alvo):** Classificação do nível de obesidade/sobrepeso do paciente.

A base de dados conta com **2.111 registros** sem valores nulos, divididos de forma equilibrada entre as classes, o que permitirá o treinamento de algoritmos multiclasse e o desenvolvimento de um painel analítico com foco em saúde preventiva.

### **Bibliotecas necessárias para análise de dados, visualização de gráficos e modelo de Machine Learning**

In [111]:
# ==============================================================================
#  1 - Importação de Bibliotecas
# ==============================================================================

# Manipulação e Análise de Dados
import pandas as pd
import numpy as np

# Visualização de Dados
import matplotlib.pyplot as plt
import seaborn as sns

# Pré-processamento e Métricas
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

# Modelos de Machine Learning
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix, ConfusionMatrixDisplay
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from sklearn.ensemble import RandomForestClassifier
import xgboost as xgb

# Métricas de validação
from sklearn.metrics import accuracy_score

# Alertas
import warnings

from sklearn.linear_model import LogisticRegression
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

In [112]:
# ==============================================================================
# 3. Carga de Dados e Configuração dos Gráficos
# ==============================================================================

df = pd.read_csv('Obesity.csv')
# fonte https://colocar aqui o link do arquivo que ficará no github

print(f"Base de dados carregada com sucesso! Total de registros: {df.shape[0]}")

Base de dados carregada com sucesso! Total de registros: 2111


# **2. Fase de Exploração e Visualização (Após o Carregamento)**
Análise Exploratória de Dados (EDA)
Nesta etapa, buscamos compreender a volatilidade e as tendências históricas do índice para fundamentar a escolha dos modelos.

Gráfico de Linha (Série Histórica): Plote o Último (Fechamento) ao longo do tempo. Isso ajuda a identificar ciclos de mercado.

Gráfico de Barras (Volume): Visualize o volume financeiro para detectar picos atípicos de negociação.

Histograma do Target: Verifique o balanceamento entre dias de Alta (1) e Baixa (0).

# **3. Fase de Exploração dos Dados (EDA)**

Nesta etapa, realizamos o diagnóstico inicial do dataset para entender suas dimensões, tipos primitivos de dados, resumo estatístico e a presença de possíveis valores ausentes.

In [113]:
# Exibindo a dimensão do Dataset
df.shape

(2111, 17)

In [114]:
# Exibindo os primeiros dados do dataframe
df.head()

,Gender,Age,Height,Weight,family_history,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,Obesity
0,Female,21.0,1.62,64.0,yes,no,2.0,3.0,Sometimes,no,2.0,no,0.0,1.0,no,Public_Transportation,Normal_Weight
1,Female,21.0,1.52,56.0,yes,no,3.0,3.0,Sometimes,yes,3.0,yes,3.0,0.0,Sometimes,Public_Transportation,Normal_Weight
2,Male,23.0,1.80,77.0,yes,no,2.0,3.0,Sometimes,no,2.0,no,2.0,1.0,Frequently,Public_Transportation,Normal_Weight
3,Male,27.0,1.80,87.0,no,no,3.0,3.0,Sometimes,no,2.0,no,2.0,0.0,Frequently,Walking,Overweight_Level_I
4,Male,22.0,1.78,89.8,no,no,2.0,1.0,Sometimes,no,2.0,no,0.0,0.0,Sometimes,Public_Transportation,Overweight_Level_II


In [115]:
# Exibindo os últimos dados do dataframe
df.tail()

,Gender,Age,Height,Weight,family_history,FAVC,FCVC,NCP,CAEC,SMOKE,CH2O,SCC,FAF,TUE,CALC,MTRANS,Obesity
2106,Female,20.976842,1.710730,131.408528,yes,yes,3.0,3.0,Sometimes,no,1.728139,no,1.676269,0.906247,Sometimes,Public_Transportation,Obesity_Type_III
2107,Female,21.982942,1.748584,133.742943,yes,yes,3.0,3.0,Sometimes,no,2.005130,no,1.341390,0.599270,Sometimes,Public_Transportation,Obesity_Type_III
2108,Female,22.524036,1.752206,133.689352,yes,yes,3.0,3.0,Sometimes,no,2.054193,no,1.414209,0.646288,Sometimes,Public_Transportation,Obesity_Type_III
2109,Female,24.361936,1.739450,133.346641,yes,yes,3.0,3.0,Sometimes,no,2.852339,no,1.139107,0.586035,Sometimes,Public_Transportation,Obesity_Type_III
2110,Female,23.664709,1.738836,133.472641,yes,yes,3.0,3.0,Sometimes,no,2.863513,no,1.026452,0.714137,Sometimes,Public_Transportation,Obesity_Type_III


In [116]:
# Exibindo as estatísticas descritivas básicas do dataframe
df.describe()

,Age,Height,Weight,FCVC,NCP,CH2O,FAF,TUE
count,2111.000000,2111.000000,2111.000000,2111.000000,2111.000000,2111.000000,2111.000000,2111.000000
mean,24.312600,1.701677,86.586058,2.419043,2.685628,2.008011,1.010298,0.657866
std,6.345968,0.093305,26.191172,0.533927,0.778039,0.612953,0.850592,0.608927
min,14.000000,1.450000,39.000000,1.000000,1.000000,1.000000,0.000000,0.000000
25%,19.947192,1.630000,65.473343,2.000000,2.658738,1.584812,0.124505,0.000000
50%,22.777890,1.700499,83.000000,2.385502,3.000000,2.000000,1.000000,0.625350
75%,26.000000,1.768464,107.430682,3.000000,3.000000,2.477420,1.666678,1.000000
max,61.000000,1.980000,173.000000,3.000000,4.000000,3.000000,3.000000,2.000000


In [117]:
# Verificando se há dados nulos no dataframe
df.isnull().sum().sum()

np.int64(0)

In [118]:
# Exibindo alguns detalhes sobre os dataframe, ex. tipagem dos dados
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2111 entries, 0 to 2110
Data columns (total 17 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Gender          2111 non-null   object 
 1   Age             2111 non-null   float64
 2   Height          2111 non-null   float64
 3   Weight          2111 non-null   float64
 4   family_history  2111 non-null   object 
 5   FAVC            2111 non-null   object 
 6   FCVC            2111 non-null   float64
 7   NCP             2111 non-null   float64
 8   CAEC            2111 non-null   object 
 9   SMOKE           2111 non-null   object 
 10  CH2O            2111 non-null   float64
 11  SCC             2111 non-null   object 
 12  FAF             2111 non-null   float64
 13  TUE             2111 non-null   float64
 14  CALC            2111 non-null   object 
 15  MTRANS          2111 non-null   object 
 16  Obesity         2111 non-null   object 
dtypes: float64(8), object(9)
memory u

# **4. Tratamento de Dados e Tradução (pt_BR)**

Para adequar o projeto à realidade do hospital e facilitar a interpretação tanto da equipe médica quanto dos analistas, realizamos a tradução completa dos nomes das colunas e categorizamos os registros textuais internos para o português padrão (pt_BR).

In [119]:
# ==============================================================================
# Dicionários de Tradução (Colunas e Dados)
# ==============================================================================

# Mapeamento de Colunas
colunas_pt = {
    'Gender': 'genero', 'Age': 'idade', 'Height': 'altura', 'Weight': 'peso',
    'family_history': 'historico_familiar_sobrepeso', 'FAVC': 'consome_alimentos_alto_calorico',
    'FCVC': 'frequencia_consumo_vegetais', 'NCP': 'numero_refeicoes_principais',
    'CAEC': 'consome_entre_refeicoes', 'SMOKE': 'fumante', 'CH2O': 'consumo_agua_diario',
    'SCC': 'monitora_calorias_diarias', 'FAF': 'frequencia_atividade_fisica',
    'TUE': 'tempo_dispositivos_tecnologicos', 'CALC': 'frequencia_consumo_alcool',
    'MTRANS': 'meio_transporte', 'Obesity': 'nivel_obesidade'
}

# Mapeamento de Conteúdos Internos (Valores das Linhas)
traducao_valores = {
    'genero': {'Female': 'Feminino', 'Male': 'Masculino'},
    'historico_familiar_sobrepeso': {'yes': 'Sim', 'no': 'Não'},
    'consome_alimentos_alto_calorico': {'yes': 'Sim', 'no': 'Não'},
    'fumante': {'yes': 'Sim', 'no': 'Não'},
    'monitora_calorias_diarias': {'yes': 'Sim', 'no': 'Não'},
    'consome_entre_refeicoes': {'no': 'Nunca', 'Sometimes': 'Às vezes', 'Frequently': 'Frequentemente', 'Always': 'Sempre'},
    'frequencia_consumo_alcool': {'no': 'Nunca', 'Sometimes': 'Às vezes', 'Frequently': 'Frequentemente', 'Always': 'Sempre'},
    'meio_transporte': {
        'Public_Transportation': 'Transporte Público',
        'Walking': 'Andando',
        'Automobile': 'Automóvel',
        'Motorbike': 'Motocicleta',
        'Bike': 'Bicicleta'
    },
    'nivel_obesidade': {
        'Insufficient_Weight': 'Abaixo do Peso',
        'Normal_Weight': 'Peso Normal',
        'Overweight_Level_I': 'Sobrepeso Grau I',
        'Overweight_Level_II': 'Sobrepeso Grau II',
        'Obesity_Type_I': 'Obesidade Grau I',
        'Obesity_Type_II': 'Obesidade Grau II',
        'Obesity_Type_III': 'Obesidade Grau III'
    }
}

# Aplicando as transformações no DataFrame
df_tratado = df.rename(columns=colunas_pt)

for coluna, mapeamento in traducao_valores.items():
    if coluna in df_tratado.columns:
        df_tratado[coluna] = df_tratado[coluna].map(mapeamento).fillna(df_tratado[coluna])

# Exibir o resultado pós-tradução
#print("Amostra dos Dados Após Padronização pt_BR")
#display(df_tratado.head())

In [120]:
# Exibir o resultado pós-tradução
print("Amostra dos Dados Após Padronização pt_BR")
display(df_tratado.head())

Amostra dos Dados Após Padronização pt_BR


,genero,idade,altura,peso,historico_familiar_sobrepeso,consome_alimentos_alto_calorico,frequencia_consumo_vegetais,numero_refeicoes_principais,consome_entre_refeicoes,fumante,consumo_agua_diario,monitora_calorias_diarias,frequencia_atividade_fisica,tempo_dispositivos_tecnologicos,frequencia_consumo_alcool,meio_transporte,nivel_obesidade
0,Feminino,21.0,1.62,64.0,Sim,Não,2.0,3.0,Às vezes,Não,2.0,Não,0.0,1.0,Nunca,Transporte Público,Peso Normal
1,Feminino,21.0,1.52,56.0,Sim,Não,3.0,3.0,Às vezes,Sim,3.0,Sim,3.0,0.0,Às vezes,Transporte Público,Peso Normal
2,Masculino,23.0,1.80,77.0,Sim,Não,2.0,3.0,Às vezes,Não,2.0,Não,2.0,1.0,Frequentemente,Transporte Público,Peso Normal
3,Masculino,27.0,1.80,87.0,Não,Não,3.0,3.0,Às vezes,Não,2.0,Não,2.0,0.0,Frequentemente,Andando,Sobrepeso Grau I
4,Masculino,22.0,1.78,89.8,Não,Não,2.0,1.0,Às vezes,Não,2.0,Não,0.0,0.0,Às vezes,Transporte Público,Sobrepeso Grau II


In [121]:
# ==============================================================================
# Importação da biblioteca Plotly
# ==============================================================================
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio

# Define o renderizador padrão para o Colab
pio.renderers.default = "colab"

# Criamos a ordem das categorias para garantir a ordenação lógica nos eixos do Plotly
ordem_obesidade = list(traducao_valores['nivel_obesidade'].values())
print("Plotly configurado!")

Plotly configurado!


# **5. Gráficos**


## Target: Distribuição do Nível de Obesidade (Gráfico de Barras)

**Texto explicativo**
Por que ele é vital? Antes de tentar prever qualquer coisa, precisamos saber se a base está desbalanceada. Se o hospital tivesse 90% de pacientes com "Peso Normal", o modelo seria preguiçoso e chutaria sempre a mesma classe.

Como vimos no diagnóstico inicial, a base é sinteticamente balanceada. Mostrar isso prova para o avaliador que você validou a distribuição da sua variável alvo.

In [122]:
# Gráfico 1: Distribuição da Variável Alvo
fig1 = px.histogram(
    df_tratado,
    x='nivel_obesidade',
    category_orders={'nivel_obesidade': ordem_obesidade},
    color='nivel_obesidade',
    color_discrete_sequence=px.colors.qualitative.Dark24,
    title='<b>Distribuição dos Pacientes por Nível de Obesidade (Base Balanceada)</b>'
)

fig1.update_layout(
    xaxis_title='Nível de Obesidade',
    yaxis_title='Quantidade de Pacientes',
    showlegend=False,
    title_x=0.5, # Centraliza o título
    height=500
)

fig1.show()

## O Peso da Genética: Histórico Familiar vs. Nível de Obesidade

**Texto Explicativo**

O Insight Clínico: A obesidade é multifatorial, mas qual o peso real da genética nesta base? Este gráfico vai mostrar de forma chocante para a equipe médica que, nos níveis de Obesidade Grau I, II e III, a quantidade de pessoas sem histórico familiar é praticamente nula

In [123]:
# Gráfico 2: Cruzamento de Histórico Familiar (Interativo)
fig2 = px.histogram(
    df_tratado,
    x='nivel_obesidade',
    color='historico_familiar_sobrepeso',
    barmode='group',
    category_orders={'nivel_obesidade': ordem_obesidade},
    color_discrete_map={'Sim': '#1f77b4', 'Não': '#aec7e8'},
    title='<b>O Fator Genético: Histórico Familiar de Sobrepeso por Classe</b>'
)

fig2.update_layout(
    xaxis_title='Nível de Obesidade',
    yaxis_title='Contagem de Pacientes',
    legend_title='Histórico Familiar?',
    title_x=0.5,
    height=500
)

fig2.show()

## O Paradoxo dos Hábitos: Atividade Física e Consumo de Água

**Texto Explicativo**
O Insight Clínico: Médicos costumam associar o ganho de peso diretamente ao sedentarismo e à baixa ingestão de água. Mas será que os dados brutos mostram uma queda linear dessas variáveis conforme a obesidade aumenta? Os boxplots revelam a distribuição (mediana, quartis e outliers) dessas variáveis contínuas em cada categoria.

In [124]:
# Gráfico 3: Boxplot Interativo de Atividade Física
fig3 = px.box(
    df_tratado,
    x='nivel_obesidade',
    y='frequencia_atividade_fisica',
    color='nivel_obesidade',
    category_orders={'nivel_obesidade': ordem_obesidade},
    title='<b>Distribuição de Frequência de Atividade Física por Classe Médica</b>',
    color_discrete_sequence=px.colors.sequential.Blues_r
)

fig3.update_layout(
    xaxis_title='Nível de Obesidade',
    yaxis_title='Frequência de Atividade Física',
    showlegend=False,
    title_x=0.5,
    height=500
)

fig3.show()

## Comportamento de Risco: Alimentos Calóricos vs. Lanchar entre Refeições (Mosaico / Heatmap de Frequência)

**Texto Explicativo**
O Insight de Negócio: Quem desenvolve obesidade severa come mais alimentos calóricos com frequência (consome_alimentos_alto_calorico) ou o perigo real está nas "beliscadas" ao longo do dia (consome_entre_refeicoes)? Cruzar essas duas variáveis categóricas usando uma matriz de calor (Heatmap) revela o cruzamento de hábitos mais perigoso da base

In [125]:
# Gráfico 4: Matriz de Hábitos Alimentares de Risco
matriz_habitos = pd.crosstab(
    df_tratado['consome_entre_refeicoes'],
    df_tratado['consome_alimentos_alto_calorico'],
    normalize='index'
) * 100

fig4 = px.imshow(
    matriz_habitos,
    text_auto='.1f',
    color_continuous_scale='YlOrRd',
    labels=dict(x="Consome Alimentos Altamente Calóricos?", y="Consome Alimentos Entre as Refeições?", color="% de Pacientes"),
    title='<b>Cruzamento Comportamental: Beliscar entre Refeições vs. Alimentos Calóricos</b>'
)

fig4.update_layout(title_x=0.5, height=500)
fig4.show()

In [126]:
# ==============================================================================
# 5.2 Análise de Correlação das Variáveis
# ==============================================================================
from sklearn.preprocessing import LabelEncoder

# Criando uma cópia para não afetar o DataFrame original de treino
df_correlacao = df_tratado.copy()

# Convertendo variáveis categóricas em numéricas para o cálculo da correlação
le_corr = LabelEncoder()
for col in df_correlacao.columns:
    if df_correlacao[col].dtype == 'object' or df_correlacao[col].dtype.name == 'category':
        df_correlacao[col] = le_corr.fit_transform(df_correlacao[col].astype(str))

# Calculando a matriz de correlação de Pearson
matriz_corr = df_correlacao.corr()

# Gerando o Heatmap interativo com Plotly
fig_corr = px.imshow(
    matriz_corr,
    text_auto='.2f',
    color_continuous_scale='RdBu_r', # Azul para positiva, Vermelho para negativa
    zmin=-1, zmax=1,
    title='<b>Matriz de Correlação de Pearson</b>'
)

fig_corr.update_layout(title_x=0.5, height=800, width=800)
fig_corr.show()

Informações adicionais dos gráficos anteriores
Você mostra a visão macro da base (Gráfico 1).
Você valida a hipótese clínica sobre genética (Gráfico 2).
Você investiga o comportamento de variáveis contínuas complexas (Gráfico 3).
Você cruza hábitos comportamentais para gerar estratégia de prevenção para o hospital (Gráfico 4)

# **7. Estratégia de Engenharia de Atributos**

Para evitar o **Vazemento de dados** (Data Leakage), realizamos primeiro a separação do dataset entre variáveis preditoras (X) e variável alvo (y), seguida da divisão em conjuntos de Treino e Teste.

As transformações categóricas e numéricas serão estruturadas através de um **ColumnTransformer** para serem integradas diretamente ao Pipeline de Machine Learning.

In [127]:
# Separação das Variáveis e Split

from sklearn.model_selection import train_test_split

# 1. Separação em Preditoras (X) e Alvo (y)
X = df_tratado.drop(columns=['nivel_obesidade'])
y = df_tratado['nivel_obesidade']

# 2. Divisão em Treino (80%) e Teste (20%) com estratificação para manter as classes balanceadas
X_treino, X_teste, y_treino, y_teste = train_test_split(
    X, y,
    test_size=0.20,
    random_state=42,
    stratify=y
)

print(f"Dados de Treino: {X_treino.shape[0]} registros")
print(f"Dados de Teste: {X_teste.shape[0]} registros")

Dados de Treino: 1688 registros
Dados de Teste: 423 registros


In [128]:
# Definição da Estrutura de Feature Engineering

from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder

# Identificando as colunas por tipo de tratamento
colunas_numericas = ['idade', 'altura', 'peso', 'frequencia_consumo_vegetais',
                     'numero_refeicoes_principais', 'consumo_agua_diario',
                     'frequencia_atividade_fisica', 'tempo_dispositivos_tecnologicos']

colunas_categoricas_nominais = ['genero', 'historico_familiar_sobrepeso',
                                'consome_alimentos_alto_calorico', 'fumante',
                                'monitora_calorias_diarias', 'meio_transporte']

# Variáveis categóricas ordinais (que possuem uma ordem lógica de intensidade)
colunas_categoricas_ordinais = ['consome_entre_refeicoes', 'frequencia_consumo_alcool']
ordem_categorias = [
    ['Nunca', 'Às vezes', 'Frequentemente', 'Sempre'], # Ordem para consome_entre_refeicoes
    ['Nunca', 'Às vezes', 'Frequentemente', 'Sempre']  # Ordem para frequencia_consumo_alcool
]

# Criando o transformador que será injetado no Pipeline
preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), colunas_numericas),
        ('cat_nominal', OneHotEncoder(drop='first', handle_unknown='ignore'), colunas_categoricas_nominais),
        ('cat_ordinal', OrdinalEncoder(categories=ordem_categorias), colunas_categoricas_ordinais)
    ]
)

print("Estrutura de Feature Engineering criada com sucesso!")

Estrutura de Feature Engineering criada com sucesso!


# **8. Treinamento e Comparação de Modelos**

Nesta etapa, estabelecemos um modelo baseline utilizando **Regressão Logística** e o comparamos com algoritmos mais complexos (**Random Forest** e **XGBoost**).
Para garantir a integridade dos testes, todos os modelos serão validados no conjunto de teste utilizando as métricas de **Acurácia** (Meta > 75%) e **F1-Score Macro** (para avaliar o desempenho balanceado entre as 7 classes de diagnóstico).

In [129]:
# Instalação do XGBoost e Importação das Métricas

from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier
from sklearn.preprocessing import LabelEncoder

# Importante: O XGBoost exige que as classes (y) sejam numéricas (0 a 6)
# Vamos codificar a variável alvo apenas para o treino/teste dos modelos
label_encoder = LabelEncoder()
y_treino_encoded = label_encoder.fit_transform(y_treino)
y_teste_encoded = label_encoder.transform(y_teste)

# Definindo o dicionário com os modelos que vamos testar
modelos = {
    'Regressão Logística (Baseline)': LogisticRegression(max_iter=1000, random_state=42),
    'Random Forest': RandomForestClassifier(random_state=42, n_estimators=100),
    'XGBoost': XGBClassifier(random_state=42, eval_metric='mlogloss')
}

# Lista para armazenar o dicionário de resultados
resultados_lista = []

print("Modelos e métricas configurados")

Modelos e métricas configurados


## **8.1 Cenário 1: Treinamento dos Modelos com Peso e Altura**

Para avaliar a capacidade do modelo em antecipar o risco de obesidade utilizando apenas o histórico familiar e os hábitos cotidianos do paciente, removemos utilizando as variáveis preditoras diretas de massa física (`peso` e `altura`) e reavaliamos os algoritmos.

In [130]:
# ==============================================================================
# Treinamento e Avaliação
# ==============================================================================

# Loop dos modelos
for nome, modelo in modelos.items():
    # Criando o Pipeline completo: Preprocessamento + Algoritmo
    pipeline_completo = Pipeline(steps=[
        ('preprocessor', preprocessor),
        ('classifier', modelo)
    ])

    # Treinando o pipeline (fit aprende os tratamentos e o modelo APENAS no treino)
    pipeline_completo.fit(X_treino, y_treino_encoded)

    # Predizendo no conjunto de teste
    y_pred = pipeline_completo.predict(X_teste)

    # Calculando as métricas
    acuracia = accuracy_score(y_teste_encoded, y_pred)
    f1 = f1_score(y_teste_encoded, y_pred, average='macro')

    # Salvando os resultados
    resultados_lista.append({
        'Modelo': nome,
        'Acurácia (Teste)': acuracia,
        'F1-Score (Macro)': f1
    })

# Convertendo a lista de resultados em um DataFrame para exibição elegante
df_resultados = pd.DataFrame(resultados_lista)

print("=== Tabela Comparativa de Performance ===")
display(df_resultados.sort_values(by='Acurácia (Teste)', ascending=False))

=== Tabela Comparativa de Performance ===


,Modelo,Acurácia (Teste),F1-Score (Macro)
2,XGBoost,0.959811,0.959259
1,Random Forest,0.936170,0.935538
0,Regressão Logística (Baseline),0.886525,0.880819


In [131]:
# ==============================================================================
# Treinamento e Avaliação dos Modelos (Formato de Texto - Padrão Fase 2)
# ==============================================================================
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, f1_score
from sklearn.preprocessing import LabelEncoder

# Importação dos algoritmos solicitados
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

# 1. O XGBoost exige que as classes sejam numéricas (0 a 6).
# Vamos codificar a variável alvo de treino e teste.
label_encoder = LabelEncoder()
y_treino_encoded = label_encoder.fit_transform(y_treino)
y_teste_encoded = label_encoder.transform(y_teste)

print("=== INICIANDO AVALIAÇÃO DOS MODELOS ===\n")

# --- MODELO 1: Regressão Logística (Baseline) ---
pipe_logistic = Pipeline(steps=[('preprocessor', preprocessor), ('classifier', LogisticRegression(max_iter=1000, random_state=42))])
pipe_logistic.fit(X_treino, y_treino_encoded)
y_pred_logistic = pipe_logistic.predict(X_teste)
print(f"Acurácia - Regressão Logística (Base Line): {accuracy_score(y_teste_encoded, y_pred_logistic):.2%}")

# --- MODELO 2: KNN ---
pipe_knn = Pipeline(steps=[('preprocessor', preprocessor), ('classifier', KNeighborsClassifier())])
pipe_knn.fit(X_treino, y_treino_encoded)
y_pred_knn = pipe_knn.predict(X_teste)
print(f"Acurácia - KNN: {accuracy_score(y_teste_encoded, y_pred_knn):.2%}")

# --- MODELO 3: SVM ---
pipe_svm = Pipeline(steps=[('preprocessor', preprocessor), ('classifier', SVC(random_state=42))])
pipe_svm.fit(X_treino, y_treino_encoded)
y_pred_svm = pipe_svm.predict(X_teste)
print(f"Acurácia - SVM: {accuracy_score(y_teste_encoded, y_pred_svm):.2%}")

# --- MODELO 4: Random Forest ---
pipe_rf = Pipeline(steps=[('preprocessor', preprocessor), ('classifier', RandomForestClassifier(random_state=42))])
pipe_rf.fit(X_treino, y_treino_encoded)
y_pred_rf = pipe_rf.predict(X_teste)
print(f"Acurácia - Random Forest: {accuracy_score(y_teste_encoded, y_pred_rf):.2%}")

# --- MODELO 5: XGBoost ---
pipe_xgb = Pipeline(steps=[('preprocessor', preprocessor), ('classifier', XGBClassifier(random_state=42, eval_metric='mlogloss'))])
pipe_xgb.fit(X_treino, y_treino_encoded)
y_pred_xgb = pipe_xgb.predict(X_teste)
print(f"Acurácia - XGBoost: {accuracy_score(y_teste_encoded, y_pred_xgb):.2%}")

print("\n" + "="*50 + "\n Evaluation concluída!")

=== INICIANDO AVALIAÇÃO DOS MODELOS ===

Acurácia - Regressão Logística (Base Line): 88.65%
Acurácia - KNN: 83.22%
Acurácia - SVM: 91.25%
Acurácia - Random Forest: 93.62%
Acurácia - XGBoost: 95.98%

 Evaluation concluída!


## **8.2 Cenário 2: Treinamento dos Modelos sem Peso e Altura (Foco Comportamental)**

Para avaliar a capacidade do modelo em antecipar o risco de obesidade utilizando apenas o histórico familiar e os hábitos cotidianos do paciente, removemos as variáveis preditoras diretas de massa física (`peso` e `altura`) e reavaliamos os algoritmos.

In [132]:
# ==============================================================================
# Preparação dos Dados e Pipeline sem Peso e Altura
# ==============================================================================
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.model_selection import train_test_split

# 1. Removendo peso e altura das variáveis preditoras
X_comportamental = df_tratado.drop(columns=['nivel_obesidade', 'peso', 'altura'])
y_comportamental = df_tratado['nivel_obesidade']

# 2. Divisão em Treino e Teste (Mantendo a mesma semente 42 para comparação justa)
X_treino_c, X_teste_c, y_treino_c, y_teste_c = train_test_split(
    X_comportamental, y_comportamental,
    test_size=0.20,
    random_state=42,
    stratify=y_comportamental
)

# Encodando a variável alvo
le_c = LabelEncoder()
y_treino_c_encoded = le_c.fit_transform(y_treino_c)
y_teste_c_encoded = le_c.transform(y_teste_c)

# 3. Atualizando o Pré-processador (Removendo 'peso' e 'altura' da lista numérica)
colunas_numericas_c = ['idade', 'frequencia_consumo_vegetais', 'numero_refeicoes_principais',
                       'consumo_agua_diario', 'frequencia_atividade_fisica', 'tempo_dispositivos_tecnologicos']

preprocessor_c = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), colunas_numericas_c),
        ('cat_nominal', OneHotEncoder(drop='first', handle_unknown='ignore'), colunas_categoricas_nominais), # Mantém as mesmas nominais
        ('cat_ordinal', OrdinalEncoder(categories=ordem_categorias), colunas_categoricas_ordinais)       # Mantém as mesmas ordinais
    ]
)

print("Novo conjunto comportamental estruturado com sucesso!")

Novo conjunto comportamental estruturado com sucesso!


In [133]:
# ==============================================================================
# 8.2 Torneio de Modelos Sem Peso e Altura (Exibição em Texto - Padrão Fase 2)
# ==============================================================================
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier

print("=== INICIANDO AVALIAÇÃO DOS MODELOS (SEM PESO E ALTURA) ===\n")

# --- MODELO 1: Regressão Logística ---
pipe_logistic_c = Pipeline(steps=[('preprocessor', preprocessor_c), ('classifier', LogisticRegression(max_iter=1000, random_state=42))])
pipe_logistic_c.fit(X_treino_c, y_treino_c_encoded)
y_pred_logistic_c = pipe_logistic_c.predict(X_teste_c)
print(f"Acurácia - Regressão Logística (Base Line): {accuracy_score(y_teste_c_encoded, y_pred_logistic_c):.2%}")

# --- MODELO 2: KNN ---
pipe_knn_c = Pipeline(steps=[('preprocessor', preprocessor_c), ('classifier', KNeighborsClassifier())])
pipe_knn_c.fit(X_treino_c, y_treino_c_encoded)
y_pred_knn_c = pipe_knn_c.predict(X_teste_c)
print(f"Acurácia - KNN: {accuracy_score(y_teste_c_encoded, y_pred_knn_c):.2%}")

# --- MODELO 3: SVM ---
pipe_svm_c = Pipeline(steps=[('preprocessor', preprocessor_c), ('classifier', SVC(random_state=42))])
pipe_svm_c.fit(X_treino_c, y_treino_c_encoded)
y_pred_svm_c = pipe_svm_c.predict(X_teste_c)
print(f"Acurácia - SVM: {accuracy_score(y_teste_c_encoded, y_pred_svm_c):.2%}")

# --- MODELO 4: Random Forest ---
pipe_rf_c = Pipeline(steps=[('preprocessor', preprocessor_c), ('classifier', RandomForestClassifier(random_state=42))])
pipe_rf_c.fit(X_treino_c, y_treino_c_encoded)
y_pred_rf_c = pipe_rf_c.predict(X_teste_c)
print(f"Acurácia - Random Forest: {accuracy_score(y_teste_c_encoded, y_pred_rf_c):.2%}")

# --- MODELO 5: XGBoost ---
pipe_xgb_c = Pipeline(steps=[('preprocessor', preprocessor_c), ('classifier', XGBClassifier(random_state=42, eval_metric='mlogloss'))])
pipe_xgb_c.fit(X_treino_c, y_treino_c_encoded)
y_pred_xgb_c = pipe_xgb_c.predict(X_teste_c)
print(f"Acurácia - XGBoost: {accuracy_score(y_teste_c_encoded, y_pred_xgb_c):.2%}")

print("\n" + "="*50 + "\n Nova avaliação concluída!")

=== INICIANDO AVALIAÇÃO DOS MODELOS (SEM PESO E ALTURA) ===

Acurácia - Regressão Logística (Base Line): 57.68%
Acurácia - KNN: 72.81%
Acurácia - SVM: 76.36%
Acurácia - Random Forest: 85.58%
Acurácia - XGBoost: 84.40%

 Nova avaliação concluída!


In [134]:
# ==============================================================================
# 9. COMPARAÇÃO FINAL ENTRE OS CENÁRIOS (Padrão de Texto Fase 2)
# ==============================================================================
print("="*60)
print("     CENÁRIO 1: MODELAGEM CLÍNICA COMPLETA (COM PESO E ALTURA)")
print("="*60)
print(f"Acurácia - Regressão Logística (Base Line): {accuracy_score(y_teste_encoded, pipe_logistic.predict(X_teste)):.2%}")
print(f"Acurácia - KNN: {accuracy_score(y_teste_encoded, pipe_knn.predict(X_teste)):.2%}")
print(f"Acurácia - SVM: {accuracy_score(y_teste_encoded, pipe_svm.predict(X_teste)):.2%}")
print(f"Acurácia - Random Forest: {accuracy_score(y_teste_encoded, pipe_rf.predict(X_teste)):.2%}")
print(f"Acurácia - XGBoost: {accuracy_score(y_teste_encoded, pipe_xgb.predict(X_teste)):.2%}")

print("\n" + "="*60)
print("     CENÁRIO 2: MODELAGEM COMPORTAMENTAL (SEM PESO E ALTURA)")
print("="*60)
print(f"Acurácia - Regressão Logística (Base Line): {accuracy_score(y_teste_c_encoded, pipe_logistic_c.predict(X_teste_c)):.2%}")
print(f"Acurácia - KNN: {accuracy_score(y_teste_c_encoded, pipe_knn_c.predict(X_teste_c)):.2%}")
print(f"Acurácia - SVM: {accuracy_score(y_teste_c_encoded, pipe_svm_c.predict(X_teste_c)):.2%}")
print(f"Acurácia - Random Forest: {accuracy_score(y_teste_c_encoded, pipe_rf_c.predict(X_teste_c)):.2%}")
print(f"Acurácia - XGBoost: {accuracy_score(y_teste_c_encoded, pipe_xgb_c.predict(X_teste_c)):.2%}")
print("="*60)

     CENÁRIO 1: MODELAGEM CLÍNICA COMPLETA (COM PESO E ALTURA)
Acurácia - Regressão Logística (Base Line): 88.65%
Acurácia - KNN: 83.22%
Acurácia - SVM: 91.25%
Acurácia - Random Forest: 93.62%
Acurácia - XGBoost: 95.98%

     CENÁRIO 2: MODELAGEM COMPORTAMENTAL (SEM PESO E ALTURA)
Acurácia - Regressão Logística (Base Line): 57.68%
Acurácia - KNN: 72.81%
Acurácia - SVM: 76.36%
Acurácia - Random Forest: 85.58%
Acurácia - XGBoost: 84.40%


# **9. Conclusão e Próximos Passos (Escolha do Modelo para Deploy)**

### **Análise Estratégica dos Resultados**
A avaliação comparativa entre os dois cenários trouxe respostas fundamentais para a viabilidade do projeto:

1. **No Cenário 1 (Clínico Completo),** os algoritmos baseados em árvores de decisão demonstraram uma performance excepcional, com o **XGBoost atingindo a acurácia máxima de 95.98%**. Esse resultado se deve à capacidade do modelo de mapear e extrair a relação não linear direta que o Peso e a Altura possuem na definição biológica do IMC e das faixas de sobrepeso.
2. **No Cenário 2 (Comportamental),** ao removermos os dados diretos de massa física, os modelos foram desafiados a identificar padrões puramente baseados em estilo de vida e genética. Embora a acurácia geral sofra uma redução natural nesse cenário mitigado, os algoritmos ainda conseguiram manter uma capacidade preditiva sólida.

### **Decisão de Negócio e Engenharia de Produção**
Com base nos requisitos do hospital e visando o máximo suporte à decisão médica em ambiente de triagem, o **XGBoost do Cenário 1** foi o vencedor incontestável. Ele não apenas superou com ampla margem a meta técnica de **75% de assertividade exigida pelo desafio**, como estabeleceu um patamar de **95.98% de precisão** para diagnósticos preditivos robustos.

**Próximo Passo:** Este pipeline campeão (`pipe_xgb`) foi serializado e exportado em formato binário (`.pkl`). Ele servirá como o motor preditivo principal da nossa aplicação web interativa desenvolvida em **Streamlit**, oferecendo aos médicos uma interface rápida, segura e de alta confiabilidade para o diagnóstico preventivo da obesidade.

In [135]:
# ==============================================================================
# 11. EXPORTAÇÃO DOS ARTEFATOS PAIOL (Criação dos arquivos .pkl)
# ==============================================================================
import pickle
import os
from google.colab import files

print("=== Iniciando a serialização dos modelos ===")

# 1. Criando a pasta 'modelos' localmente para manter a organização do GitHub
os.makedirs('modelos', exist_ok=True)

# 2. Salvando o Modelo Campeão (Pipeline do XGBoost do Cenário 1)
caminho_modelo = 'modelos/modelo_obesidade_xgb.pkl'
with open(caminho_modelo, 'wb') as arquivo_modelo:
    pickle.dump(pipe_xgb, arquivo_modelo)
print(f"✓ Arquivo '{caminho_modelo}' gerado com sucesso!")

# 3. Salvando o LabelEncoder (Decodificador de Rótulos Médicos)
caminho_encoder = 'modelos/label_encoder.pkl'
with open(caminho_encoder, 'wb') as arquivo_encoder:
    pickle.dump(label_encoder, arquivo_encoder)
print(f"✓ Arquivo '{caminho_encoder}' gerado com sucesso!")

print("\n=== Baixando os arquivos para o seu computador ===")
# 4. Comando do Colab para forçar o download automático dos arquivos para sua máquina
files.download(caminho_modelo)
files.download(caminho_encoder)

print("Verifique a sua pasta de 'Downloads' no computador!")

=== Iniciando a serialização dos modelos ===
✓ Arquivo 'modelos/modelo_obesidade_xgb.pkl' gerado com sucesso!
✓ Arquivo 'modelos/label_encoder.pkl' gerado com sucesso!

=== Baixando os arquivos para o seu computador ===


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Verifique a sua pasta de 'Downloads' no computador!


In [136]:
# !pip install streamlit

In [137]:
import sklearn
import xgboost
import pandas
import numpy
import streamlit as st

print("Copie estas versões para o seu requirements.txt:\n")
print(f"sklearn=={sklearn.__version__}")
print(f"pandas=={pandas.__version__}")
print(f"numpy=={numpy.__version__}")
print(f"scikit-learn=={sklearn.__version__}")
print(f"xgboost=={xgboost.__version__}")
print("streamlit==1.32.0") # Versão estável recomendada

Copie estas versões para o seu requirements.txt:

sklearn==1.6.1
pandas==2.2.2
numpy==2.0.2
scikit-learn==1.6.1
xgboost==3.2.0
streamlit==1.32.0
